src/workflows-app/RunDetails.ts

```diff
@@ -0,0 +1,472 @@
+  const { submissionId, workflowId } = props;
+
+  /* State Setup */
+
+  /* Leo app state of Cromwell, which includes both the app status and the app URL.
+   Must be retrieved from Leo, and is necessary for all Cromwell API calls.
+   Cached here so we only fetch it once. */
+  const [cromwellProxyState, setCromwellProxyState] = useState<CromwellProxyUrlState>();
+
+  /* Workflow metadta, which includes the call objects array, among other things. 
+  The web request to fetch this data may be slow. */
+  const [workflowMetadata, setWorkflowMetadata] = useState<WorkflowMetadata>();
+  const [callObjects, setCallObjects] = useState<any[] | undefined>([]);
+
+  const [costMetadata, setCostMetadata] = useState<WorkflowMetadata>();
+  const [isCostMetadataLoading, setIsCostMetadataLoading] = useState<boolean>(true);
+
+  const [rootWorkflowCost, setRootWorkflowCost] = useState<number>();
+  /* If a workflow fails, we make a special request to get the failed tasks. */
+  const [failedTasks, setFailedTasks] = useState<any[]>([]);
+
+  /* Modal states. Child components are given a callback to open these modals. */
+  const [logModalState, setLogModalState] = useState<LogModalState>({
+    showing: false,
+    title: '',
+    logsArray: [],
+    templateTesLog: '',
+  });
+
+  const [taskModalState, setTaskModalState] = useState<TaskModalState>({
+    showing: false,
+    taskDataTile: '',
+    taskDataJson: {},
+  });
+
+  /* State to track if the workflow failed to load. */
+  const [loadWorkflowFailed, setLoadWorkflowFailed] = useState(false);
+  const signal = useCancellation();
+  const stateRefreshTimer = useRef<NodeJS.Timeout>();
+  const { captureEvent } = useMetricsEvent();
+
+  const [sasToken, setSasToken] = useState('');
+
+  const metadataArchiveStatus = useMemo(() => workflowMetadata?.metadataArchiveStatus, [workflowMetadata]);
+
+  /* Callback for child components to use to open the log modal */
+  const showLogModal = useCallback((modalTitle, logsArray, tesLog) => {
+    setLogModalState({
+      showing: true,
+      title: modalTitle,
+      logsArray,
+      templateTesLog: tesLog,
+    });
+  }, []);
+
+  /* Callback for child components to use to open the task data modal */
+  const showTaskDataModal = useCallback((taskDataTitle, taskJson) => {
+    setTaskModalState({
+      showing: true,
+      taskDataTile: taskDataTitle,
+      taskDataJson: taskJson,
+    });
+  }, []);
+
+  /* Data fetching */
+
+  // Fetch the Cromwell App URL and status from Leo.
+  useEffect(() => {
+    const fetchCromwellProxyState = async () => {
+      try {
+        const { cromwellProxyUrlState } = await loadAppUrls(workspaceId, 'cromwellProxyUrlState');
+        setCromwellProxyState(cromwellProxyUrlState);
+      } catch (error) {
+        notify('error', 'Error fetching Cromwell proxy URL for workspace.', {
+          detail: error instanceof Response ? await error.text() : error,
+        });
+      }
+    };
+    fetchCromwellProxyState();
```

Comment:
My understanding is that this is a fairly common pattern in react, and it confers a few advantages:

Remember: `useEffect()` will run when the component mounts *and* whenever any of its dependencies change. 

The main idea is that defining a function inside the `useEffect` lets us not list it as a dependency.
- This helps prevent infinite loop bugs where state the function uses changes, causing the function to change, causing `useEffect()` to rerun, changing state the function uses...
- If we define it outside of the component, we can't use `setCromwellProxyState`. 
- If we define it inside of the component but outside the `useEffect`, it causes the dependencies of the `useEffect` to change on every re-render, since the function is redeclared. #reactstuff. This doubles the requests. We could work around this by wrapping it in a `useCallback`.
- We get to me much more explicit + specific about what dependencies this `useEffect` has.  

Also, we get to give the function a name.

Reply:
But, yeah, we could also just call the logic directly inside of the `useEffect` and that would work just as well. I think there may be performance benefits to doing it as you suggest (by preventing function redeclarations, and therefore memory allocations on rerender), but I'm not super sure on that.

Reply:
Cool, that's a lot of good reasons to define the function inside the `useEffect`, but why define a function at all? Is there a difference between what's here and this?
```
useEffect(() => {
    (async () => {
      try {
        const { cromwellProxyUrlState } = await loadAppUrls(workspaceId, 'cromwellProxyUrlState');
        setCromwellProxyState(cromwellProxyUrlState);
      } catch (error) {
        notify('error', 'Error fetching Cromwell proxy URL for workspace.', {
          detail: error instanceof Response ? await error.text() : error,
        });
      }
    })();
}, [workspaceId]);
```
Leaving this here for posterity but while looking up the correct syntax to run async code as above, I think I answered my own question - to run a block async you need to define a function-like thing and then invoke it, and at that point (as you say) it's nice to give the function a name.

Reply:
Ah, I didn't see your second comment because I was too busy writing mine, makes sense 👍